src/vm/amd64/pinvokestubs.S

```diff
@@ -0,0 +1,277 @@
+
+NESTED_ENTRY PInvokeStubForHost, _TEXT, NoHandler
+        alloc_stack     PInvokeStubForHost_STACK_FRAME_SIZE
+        END_PROLOGUE
+
+        // spill args
+        mov             [rsp + PInvokeStubForHost_PARAM_REGISTERS_OFFSET +  0h], rdi
+        mov             [rsp + PInvokeStubForHost_PARAM_REGISTERS_OFFSET +  8h], rdx
+        mov             [rsp + PInvokeStubForHost_PARAM_REGISTERS_OFFSET + 10h], rdx
+        mov             [rsp + PInvokeStubForHost_PARAM_REGISTERS_OFFSET + 18h], rcx
+        movdqa          [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET +  0h], xmm0
+        movdqa          [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET + 10h], xmm1
+        movdqa          [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET + 20h], xmm2
+        movdqa          [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET + 30h], xmm3
+        movdqa          [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET + 40h], xmm4
+        movdqa          [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET + 50h], xmm5
+        movdqa          [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET + 60h], xmm6
+        movdqa          [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET + 70h], xmm7
+        
+        
+        // PInvokeStubForHostWorker(#stack args, stack frame, this)
+        mov             rdx, rdi
+        mov             rdi, r11
+        mov             rsi, rsp
+        call            C_FUNC(PInvokeStubForHostWorker)
+        
+        // unspill return value
+        mov             rax,  [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET +  0h]
+        movdqa          xmm0, [rsp + PInvokeStubForHost_XMM_SAVE_OFFSET + 10h]
+        
+        add             rsp, PInvokeStubForHost_STACK_FRAME_SIZE
+        ret
+NESTED_END PInvokeStubForHost, _TEXT
+
+
+PInvokeStubForHostInner_STACK_FRAME_SIZE = 0
+
+// integer registers saved in prologue
+PInvokeStubForHostInner_NUM_REG_PUSHES = 2
+PInvokeStubForHostInner_STACK_FRAME_SIZE = PInvokeStubForHostInner_STACK_FRAME_SIZE + PInvokeStubForHostInner_NUM_REG_PUSHES*8
+
+// Ensure that the new rsp will be 16-byte aligned.
+.if ((PInvokeStubForHostInner_STACK_FRAME_SIZE + 8) % 16) != 0
+PInvokeStubForHostInner_STACK_FRAME_SIZE = PInvokeStubForHostInner_STACK_FRAME_SIZE + 8
+.endif
+
+// Return address is immediately above the local variables.
+PInvokeStubForHostInner_RETURN_ADDRESS_OFFSET = PInvokeStubForHostInner_STACK_FRAME_SIZE
+PInvokeStubForHostInner_PARAM_REGISTERS_OFFSET = PInvokeStubForHostInner_RETURN_ADDRESS_OFFSET + 8
+
+PInvokeStubForHostInner_FRAME_OFFSET = PInvokeStubForHost_CALLEE_SCRATCH_SIZE
+
+// rdi - #stack args
+// rsi - PInvokeStubForHost's stack frame
+// rdx  - target address
+NESTED_ENTRY PInvokeStubForHostInner, _TEXT, NoHandler
+
+        push_nonvol_reg rbp
+        push_nonvol_reg r12
+        alloc_stack     PInvokeStubForHostInner_FRAME_OFFSET + PInvokeStubForHostInner_STACK_FRAME_SIZE - PInvokeStubForHostInner_NUM_REG_PUSHES*8
+        lea rbp, [rsp + PInvokeStubForHostInner_FRAME_OFFSET]
```

Comment:
Does the unwinder work correctly with this? I would think that we need to use the set_cfa_register macro here to ensure proper unwinding.